src/IO/H5/Python/IterElements.py

```diff
@@ -4,16 +4,62 @@
 import numpy as np
 import spectre.IO.H5 as spectre_h5
 from dataclasses import dataclass
-from spectre.Domain import ElementId
-from spectre.Spectral import Mesh
-from typing import Union, Iterable, Optional
+from spectre.DataStructures.Tensor.EagerMath import determinant
+from spectre.Domain import (ElementId, ElementMap, deserialize_domain,
+                            deserialize_functions_of_time, FunctionOfTime)
+from spectre.Domain.CoordinateMaps import (
+    CoordinateMapElementLogicalToInertial1D,
+    CoordinateMapElementLogicalToInertial2D,
+    CoordinateMapElementLogicalToInertial3D)
+from spectre.Spectral import Mesh, logical_coordinates
+from typing import Union, Iterable, Optional, Dict
+
+# functools.cached_property was added in Py 3.8. Fall back to a plain
+# `property` in Py 3.7.
+try:
+    from functools import cached_property
+except ImportError:
+    cached_property = property
 
 
 @dataclass
 class Element:
     id: Union[ElementId[1], ElementId[2], ElementId[3]]
     mesh: Union[Mesh[1], Mesh[2], Mesh[3]]
-    inertial_coords: np.ndarray
+    map: Union[CoordinateMapElementLogicalToInertial1D,
+               CoordinateMapElementLogicalToInertial2D,
+               CoordinateMapElementLogicalToInertial3D]
+    time: Optional[float]
+    functions_of_time: Optional[Dict[str, FunctionOfTime]]
+    # Offset and length in contiguous tensor data corresponding to this element
+    data_slice: slice
+
+    @property
+    def dim(self):
+        return self.mesh.dim
+
+    @cached_property
+    def logical_coordinates(self):
+        return logical_coordinates(self.mesh)
+
+    @cached_property
```

Comment:
if you mutate the `self.time`, the cache will not be aware of this afaik, so it would always give you the value of the first evaluation, no?

Reply:
The time won't be mutated because a new `Element` is returned at every observation ID. Pushed a fixup that makes the Element class immutable.

Reply:
ok, I guess that makes sense once we have h-refinement and the elements can change between time steps.